File: private.http/Objects/HttpItem.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Text;
using System.Xml.Serialization;
using Microsoft.VisualBasic.CompilerServices;



namespace @private.http
{
    [Serializable()]
    [XmlInclude(typeof(HttpItem.HttpItemApplication))]
    [XmlInclude(typeof(HttpItem.HttpItemFile))]
    [XmlInclude(typeof(HttpItem.HttpItemFolder))]
    public abstract class HttpItem
    {
        private string _VirtualPath = "";

        public string Name { get; set; } = "";
        public List<HttpItem> Content { get; set; } = new List<HttpItem>();
        public byte[] Data { get; set; }

        public string VirtualPath
        {
            get
            {
                return _VirtualPath;
            }

            set
            {
                _VirtualPath = value.Replace("//", "/");
            }
        }

        public abstract byte[] GetContent(Dictionary<string, string> Params);

        public virtual string Getmime()
        {
            return MimeTypes.MimeTypeMap.GetMimeType(".html");
        }

        public class HttpItemApplication : HttpItem
        {
            public override byte[] GetContent(Dictionary<string, string> Params)
            {
                var defdoc = this.FolderDefaultDocument(this);
                if (defdoc is object)
                    return defdoc.GetContent(Params);
                return this.FolderContentView();
            }

            public void Save(string FileName)
            {
                var serializer = new XmlSerializer(typeof(HttpItemApplication), new XmlRootAttribute("Application"));
                using (var file = System.IO.File.Open(FileName, System.IO.FileMode.Create, System.IO.FileAccess.Write))
                {
                    serializer.Serialize(file, this);
                }

                Logger.Append("Application saved.");
            }

            public static HttpItem.HttpItemApplication Load(string FileName)
            {
                var serializer = new XmlSerializer(typeof(HttpItem.HttpItemApplication), new XmlRootAttribute("Application"));
                using (var file = System.IO.File.OpenRead(FileName))
                {
                    Logger.Append("Application loaded.");
                    return (HttpItem.HttpItemApplication)serializer.Deserialize(file);
                }

                
            }
        }

        public class HttpItemFolder : HttpItem
        {
            public override byte[] GetContent(Dictionary<string, string> Params)
            {
                var defdoc = this.FolderDefaultDocument(this);
                if (defdoc is object)
                    return defdoc.Data;
                return this.FolderContentView();
            }
        }

        public class HttpItemFile : HttpItem
        {
            private Script _script;

            public object Extension
            {
                get
                {
                    if (this.Name.Contains("."))
                    {
                        return this.Name.Substring(this.Name.LastIndexOf(".")).ToLower();
                    }
                    else
                    {
                        return "";
                    }
                }
            }

            public override byte[] GetContent(Dictionary<string, string> Params)
            {
                if (Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(Extension, Script.Extension, false)))
                {
                    return ExecuteScript(Params);
                }
                else
                {
                    return this.Data;
                }
            }

            public override string Getmime()
            {
                return MimeTypes.MimeTypeMap.GetMimeType(Conversions.ToString(Extension));
            }

            public byte[] ExecuteScript(Dictionary<string, string> Parameters)
            {
                try
                {
                    string ScriptCode = Encoding.UTF8.GetString(this.Data);
                    var page = new HtmlPage();
                    page.Begin(this.VirtualPath);

                    // SAMPLE
                    // Using @private.http;
                    // Public Class Script
                    // {
                    // Public void Render(HtmlPage page)
                    // {
                    // page.Writer.AppendLine("<hr>");
                    // }
                    // }
                    if (_script is null)
                    {
                        Logger.Append($"Compiling script for '{this.VirtualPath}'");
                        _script = CSScriptLib.CSScript.Evaluator.ReferenceAssembliesFromCode(ScriptCode).ReferenceAssemblyOf<HtmlPage>().LoadCode<Script>(ScriptCode);
                    }

                    _script.Application = Program.CurrentApplication;
                    _script.Page = page;
                    _script.Current = this;
                    Logger.Append($"Executing script '{this.VirtualPath}'");
                    _script.Render(Parameters);

                    // Parameters.Values.ToArray()
                    page.Close();
                    return Encoding.UTF8.GetBytes(page.Source);
                }
                catch (Exception ex)
                {
                    Logger.Append("Error executing script: " + ex.Message);
                    return Array.Empty<byte>();
                }
            }
        }

        public HttpItem FromPath(string path)
        {
            if ((VirtualPath.ToLower() ?? "") == (path.ToLower() ?? ""))
            {
                return this;
            }
            else
            {
                foreach (HttpItem i in Content)
                {
                    var r = i.FromPath(path);
                    if (r is object)
                        return r;
                }
            }

            return null;
        }

        private HttpItem FolderDefaultDocument(HttpItem Folder)
        {
            HttpItem defdoc;
            defdoc = Content.FirstOrDefault(x => (x.Name.ToLower() ?? "") == ($"index{Script.Extension}" ?? ""));
            if (defdoc is null)
            {
                defdoc = Content.FirstOrDefault(x => x.Name.ToLower() == "index.html");
            }

            return defdoc;
        }

        private byte[] FolderContentView()
        {
            var page = new HtmlPage();
            page.Begin(VirtualPath);
            page.Writer.AppendLine($"<div>Contents of <b>{VirtualPath}</b></div>");
            page.Writer.AppendLine($"<hr>");
            foreach (HttpItem it in Content)
                page.Writer.AppendLine($"<div><a href={it.VirtualPath}>{it.Name}</a></div>");
            page.Close();
            return Encoding.UTF8.GetBytes(page.Source);
        }
    }
}

File: private.http/Objects/HttpItem.vb
Imports System.Text
Imports System.Xml.Serialization


<Serializable()>
<XmlInclude(GetType(HttpItem.HttpItemApplication))>
<XmlInclude(GetType(HttpItem.HttpItemFile))>
<XmlInclude(GetType(HttpItem.HttpItemFolder))>
Public MustInherit Class HttpItem
    Private _VirtualPath As String = ""
    Public Property Name As String = ""
    Public Property Content As New List(Of HttpItem)
    Public Property Data As Byte()

    Public Property VirtualPath As String
        Get
            Return _VirtualPath
        End Get
        Set(value As String)
            _VirtualPath = value.Replace("//", "/")
        End Set
    End Property


    Public MustOverride Function GetContent(Params As Dictionary(Of String, String)) As Byte()
    Public Overridable Function Getmime() As String
        Return MimeTypes.MimeTypeMap.GetMimeType(".html")
    End Function

    Public Class HttpItemApplication
        Inherits HttpItem

        Public Overrides Function GetContent(Params As Dictionary(Of String, String)) As Byte()
            Dim defdoc As HttpItem = FolderDefaultDocument(Me)
            If defdoc IsNot Nothing Then Return defdoc.GetContent(Params)
            Return FolderContentView()
        End Function

        Public Sub Save(FileName As String)
            Dim serializer As New XmlSerializer(GetType(HttpItemApplication), New XmlRootAttribute("Application"))
            Using file As System.IO.FileStream = System.IO.File.Open(FileName, IO.FileMode.Create, IO.FileAccess.Write)
                serializer.Serialize(file, Me)
            End Using

            Logger.Append("Application saved.")

        End Sub

        Public Shared Function Load(FileName As String) As HttpItemApplication
            Dim serializer As New XmlSerializer(GetType(HttpItemApplication), New XmlRootAttribute("Application"))
            Using file = System.IO.File.OpenRead(FileName)
                Return DirectCast(serializer.Deserialize(file), HttpItemApplication)
            End Using
            Logger.Append("Application loaded.")
        End Function
    End Class

    Public Class HttpItemFolder
        Inherits HttpItem

        Public Overrides Function GetContent(Params As Dictionary(Of String, String)) As Byte()
            Dim defdoc As HttpItem = FolderDefaultDocument(Me)
            If defdoc IsNot Nothing Then Return defdoc.Data
            Return FolderContentView()
        End Function

    End Class

    Public Class HttpItemFile
        Inherits HttpItem

        Private _script As Script
        Public ReadOnly Property Extension
            Get
                If Me.Name.Contains(".") Then
                    Return Me.Name.Substring(Me.Name.LastIndexOf(".")).ToLower
                Else
                    Return ""
                End If
            End Get
        End Property

        Public Overrides Function GetContent(Params As Dictionary(Of String, String)) As Byte()
            If Extension = Script.Extension Then
                Return ExecuteScript(Params)
            Else
                Return Me.Data
            End If
        End Function

        Public Overrides Function Getmime() As String
            Return MimeTypes.MimeTypeMap.GetMimeType(Me.Extension)
        End Function


        Public Function ExecuteScript(Parameters As Dictionary(Of String, String)) As Byte()
            Try
                Dim ScriptCode As String = Encoding.UTF8.GetString(Data)
                Dim page As New HtmlPage
                page.Begin(Me.VirtualPath)

                'SAMPLE
                'Using @private.http;
                'Public Class Script
                '{
                'Public void Render(HtmlPage page)
                '{
                'page.Writer.AppendLine("<hr>");
                '}
                '}
                If _script Is Nothing Then
                    Logger.Append($"Compiling script for '{Me.VirtualPath}'")
                    _script = CSScriptLib.CSScript.Evaluator.
                    ReferenceAssembliesFromCode(ScriptCode).
                    ReferenceAssemblyOf(Of HtmlPage)().
                    LoadCode(Of Script)(ScriptCode)
                End If
                _script.Application = Program.CurrentApplication
                _script.Page = page
                _script.Current = Me

                Logger.Append($"Executing script '{Me.VirtualPath}'")
                _script.Render(Parameters)

                ' Parameters.Values.ToArray()
                page.Close()

                Return Encoding.UTF8.GetBytes(page.Source)

            Catch ex As Exception
                Logger.Append("Error executing script: " & ex.Message)
                Return Array.Empty(Of Byte)
            End Try
        End Function
    End Class

    Public Function FromPath(path As String) As HttpItem

        If Me.VirtualPath.ToLower = path.ToLower Then
            Return Me
        Else
            For Each i As HttpItem In Me.Content
                Dim r As HttpItem = i.FromPath(path)
                If r IsNot Nothing Then Return r
            Next
        End If

        Return Nothing
    End Function

    Private Function FolderDefaultDocument(Folder As HttpItem) As HttpItem
        Dim defdoc As HttpItem
        defdoc = Me.Content.FirstOrDefault(Function(x)
                                               Return x.Name.ToLower = $"index{Script.Extension}"
                                           End Function)

        If defdoc Is Nothing Then
            defdoc = Me.Content.FirstOrDefault(Function(x)
                                                   Return x.Name.ToLower = "index.html"
                                               End Function)
        End If

        Return defdoc
    End Function
    Private Function FolderContentView() As Byte()
        Dim page As New HtmlPage
        page.Begin(Me.VirtualPath)

        page.Writer.AppendLine($"<div>Contents of <b>{Me.VirtualPath}</b></div>")
        page.Writer.AppendLine($"<hr>")
        For Each it As HttpItem In Me.Content
            page.Writer.AppendLine($"<div><a href={it.VirtualPath}>{it.Name}</a></div>")
        Next
        page.Close()

        Return Encoding.UTF8.GetBytes(page.Source)

    End Function

End Class
